src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,461 @@
+// limitations under the License.
+use super::Provide;
+use crate::authenticators::ApplicationName;
+use crate::key_id_managers::{KeyTriple, ManageKeyIDs};
+use log::{error, info};
+use parsec_interface::operations::key_attributes::*;
+use parsec_interface::operations::ProviderInfo;
+use parsec_interface::operations::{OpAsymSign, ResultAsymSign};
+use parsec_interface::operations::{OpAsymVerify, ResultAsymVerify};
+use parsec_interface::operations::{OpCreateKey, ResultCreateKey};
+use parsec_interface::operations::{OpDestroyKey, ResultDestroyKey};
+use parsec_interface::operations::{OpExportPublicKey, ResultExportPublicKey};
+use parsec_interface::operations::{OpImportKey, ResultImportKey};
+use parsec_interface::operations::{OpListOpcodes, ResultListOpcodes};
+use parsec_interface::requests::{Opcode, ProviderID, ResponseStatus, Result};
+use serde::{Deserialize, Serialize};
+use serde_asn1_der::asn1_wrapper::*;
+use std::sync::{Arc, Mutex, RwLock};
+use tss_esapi::{
+    constants::TPM2_ALG_SHA256, response_code::Tss2ResponseCodeKind, utils::AsymSchemeUnion,
+    utils::Signature, utils::TpmsContext, Tcti,
+};
+use uuid::Uuid;
+extern crate num_bigint_dig as num_bigint;
+use num_bigint::{BigInt, Sign};
+
+const SUPPORTED_OPCODES: [Opcode; 7] = [
+    Opcode::CreateKey,
+    Opcode::DestroyKey,
+    Opcode::AsymSign,
+    Opcode::AsymVerify,
+    Opcode::ImportKey,
+    Opcode::ExportPublicKey,
+    Opcode::ListOpcodes,
+];
+
+const ROOT_KEY_SIZE: usize = 2048;
+const ROOT_KEY_AUTH_SIZE: usize = 32;
+
+pub struct TpmProvider {
+    // The Mutex is needed both because interior mutability is needed to the ESAPI Context
+    // structure that is shared between threads and because two threads are not allowed the same
+    // ESAPI context simultaneously.
+    esapi_context: Mutex<tss_esapi::TransientObjectContext>,
+    // The Key ID Manager stores the key context and its associated authValue (a PasswordContext
+    // structure).
+    key_id_store: Arc<RwLock<dyn ManageKeyIDs + Send + Sync>>,
+}
+
+// Public exponent value for all RSA keys.
+const PUBLIC_EXPONENT: [u8; 3] = [0x01, 0x00, 0x01];
+const AUTH_VAL_LEN: usize = 32;
+
+// The RSA Public Key data are DER encoded with the following representation:
+// RSAPublicKey ::= SEQUENCE {
+//     modulus            INTEGER,  -- n
+//     publicExponent     INTEGER   -- e
+// }
+#[derive(Serialize, Deserialize, Debug)]
+struct RsaPublicKey {
+    modulus: IntegerAsn1,
+    public_exponent: IntegerAsn1,
+}
+
+// The PasswordContext is what is stored by the Key ID Manager.
+#[derive(Serialize, Deserialize)]
+struct PasswordContext {
+    context: TpmsContext,
+    auth_value: Vec<u8>,
+}
+
+/// Inserts a new mapping in the Key ID manager that stores the PasswordContext.
+fn insert_password_context(
+    store_handle: &mut dyn ManageKeyIDs,
+    key_triple: KeyTriple,
+    password_context: PasswordContext,
+) -> Result<()> {
+    let error_storing = |e| {
```

Comment:
Nice!!! I actually believe this could be moved to be something surfaced by the KeyIdManager - apart from the text everything seems general, maybe the KIM could do this before returning to the providers?

Reply:
Same for many of the `or_else` handlers in here

Reply:
The thing is that the `KeyIDManager` returns a `String` in its `Err` variant, so we would still need an `or_else` to convert it to `ResponseStatus` :/

Reply:
Yes, but we could do that in the manager itself

Reply:
Ok, got it now, that is a good idea!

Reply:
Will create an issue.

Reply:
In #77